File: dsa-practice-ram/dsa-practice-ram/Arrays/2D/SpiralMatrix.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace dsa_practice_ram.Arrays._2D
{
    internal class SpiralMatrix
    {
        public void PrintSpiralMatrix()
        {
            /*  
             *             left              right
             *              0   1   2   3   4   5
             *             ---------------------
             *      top 0 | 1   2   3   4   5   6
             *          1 | 20  21  22  23  24  7
             *          2 | 19  33  32  34  25  8
             *          3 | 18  31  36  35  26  9
             *          4 | 17  30  29  28  27  10
             *   bottom 5 | 16  15  14  13  12  11
             *     
             * */

            List<List<int>> spiralMatrix = new List<List<int>>
            {
                new List<int> { 1,  2,  3,  4,  5,  6 },
                new List<int> { 20, 21, 22, 23, 24, 7 },
                new List<int> { 19, 32, 33, 34, 25, 8 },
                new List<int> { 18, 31, 36, 35, 26, 9 },
                new List<int> { 17, 30, 29, 28, 27, 10 },
                new List<int> { 16, 15, 14, 13, 12, 11 }
            };

            int rows = spiralMatrix.Count;
            int cols = spiralMatrix[0].Count;

            int top = 0, bottom = rows - 1, left = 0, right = cols - 1;

            List<int> result = new();

            while (left <= right && top <= bottom)
            {
                for (int i = left; i <= right; i++)
                {
                    result.Add(spiralMatrix[top][i]);
                }

                top++;

                // This ensures that before performing a bottom-row traversal or left-column traversal, we always
                // verify the matrix is still valid.

                for (int i = top; i <= bottom; i++)
                {
                    result.Add(spiralMatrix[i][right]);
                }

                right--;

                if (top <= bottom)
                {
                    for (int i = right; i >= left; i--) // r < l
                    {
                        result.Add(spiralMatrix[bottom][i]);
                    }

                    bottom--;
                }

                if (left <= right)
                {
                    for (int i = bottom; i >= top; i--)
                    {
                        result.Add(spiralMatrix[i][left]);
                    }

                    left++;
                }
            }

            Console.WriteLine(string.Join(", ", result));
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/spiral_matrix.py
"""Spiral order traversal of a matrix."""

# C# SpiralMatrix.PrintSpiralMatrix

def print_spiral_matrix(spiral_matrix=None):
    """Return the elements of a matrix in spiral order.

    Problem: Traverse a matrix in spiral order.
    Approach: Maintain top/bottom/left/right boundaries and peel layers.
    Time complexity: O(r * c)
    Space complexity: O(r * c) for output
    """
    if spiral_matrix is None:
        spiral_matrix = [
            [1, 2, 3, 4, 5, 6],
            [20, 21, 22, 23, 24, 7],
            [19, 32, 33, 34, 25, 8],
            [18, 31, 36, 35, 26, 9],
            [17, 30, 29, 28, 27, 10],
            [16, 15, 14, 13, 12, 11],
        ]

    rows = len(spiral_matrix)
    cols = len(spiral_matrix[0]) if rows else 0
    top = 0
    bottom = rows - 1
    left = 0
    right = cols - 1

    result = []

    while left <= right and top <= bottom:
        for i in range(left, right + 1):
            result.append(spiral_matrix[top][i])
        top += 1

        for i in range(top, bottom + 1):
            result.append(spiral_matrix[i][right])
        right -= 1

        if top <= bottom:
            for i in range(right, left - 1, -1):
                result.append(spiral_matrix[bottom][i])
            bottom -= 1

        if left <= right:
            for i in range(bottom, top - 1, -1):
                result.append(spiral_matrix[i][left])
            left += 1

    return result
